src/bit_machine/mod.rs

```diff
@@ -366,66 +380,93 @@ impl BitMachine {
     }
 
     fn exec_jet<J: Jet>(&mut self, jet: J, env: &J::Environment) -> Result<(), JetFailed> {
-        use simplicity_sys::c_jets::frame_ffi::{c_readBit, c_writeBit, CFrameItem};
-        use simplicity_sys::c_jets::uword_width;
-        use simplicity_sys::ffi::UWORD;
+        use crate::ffi::c_jets::frame_ffi::{c_readBit, c_writeBit, CFrameItem};
+        use crate::ffi::c_jets::uword_width;
+        use crate::ffi::ffi::UWORD;
+
+        /// Create new C read frame that contains `bit_width` many bits from active read frame.
+        ///
+        /// Return C read frame together with underlying buffer.
+        ///
+        /// ## Panics
+        ///
+        /// Active read frame has fewer bits than `bit_width`.
+        unsafe fn get_input_frame(
```

Comment:
In b561354638232b5884f8197a47e8cfda415a16f5:

Ok, awesome, this is looking much cleaner.

Should add a `## Safety` section to the doc simply saying that it returns a `CFrameItem` which points into the returned `Vec` and that the `Vec` must outlive the `CFrameItem`.

*Alternately* we could create a new opaque type `struct BackedCFrameItem` which holds the `Vec`, implements `Deref<CFrameItem>`, and which would then ensure that you were obeying the lifetime rules. I *think* that if you did this, you could eliminate `unsafe` from this method.

Reply:
Ok, will do. I haven't been working much with unsafe, so these comments on how to write unsafe Rust (somewhat) safely are appreciated :)